app/app.go

```diff
@@ -0,0 +1,108 @@
+// It should be used by runners to determine an app's capabilities and create an instance of the app to run.
+type Provider interface {
+	// Manifest returns a Manifest, which may contain ManifestData or may point to a location where ManifestData can be fetched from.
+	// The runner should use the ManifestData to determine app capabilities.
+	Manifest() Manifest
+	// NewApp creates a new App instance using the provided config, or returns an error if an App cannot be instantiated.
+	NewApp(Config) (App, error)
+}
+
+// Runnable represents a type which can be run until it errors or the provided channel is stopped (or receives a message)
+type Runnable interface {
+	// Run runs the process and blocks until one of the following conditions are met:
+	// * An unrecoverable error occurs, in which case it returns the error
+	// * The provided channel is closed, in which case processing should stop and the method should return
+	// * The provided channel is sent a message, in which case processing should stop and the method should return
+	// * The process completes and does not need to run again
+	Run(<-chan struct{}) error
+}
+
+type AdmissionRequest resource.AdmissionRequest
+type MutatingResponse resource.MutatingResponse
+
+// App represents an app platform application logical structure.
+// An App is typically run with a wrapper, such as simple.NewStandaloneOperator,
+// which will present a runtime layer (such as kubernetes webhooks in the case of an operator),
+// and translate those into calls to the App. The wrapper is typically also responsible for lifecycle management
+// and running the Runnable provided by Runner().
+// Pre-built implementations of App exist in the simple package, but any type which implements App
+// should be capable of being run by an app wrapper.
+type App interface {
+	// Validate validates the incoming request, and returns an error if validation fails
+	Validate(ctx context.Context, request *resource.AdmissionRequest) error
+	// Mutate runs mutation on the incoming request, responding with a MutatingResponse on success, or an error on failure
+	Mutate(ctx context.Context, request *resource.AdmissionRequest) (*resource.MutatingResponse, error)
+	// Convert converts the object based on the ConversionRequest, returning a RawObject which MUST contain
+	// the converted bytes and encoding (Raw and Encoding respectively), and MAY contain the Object representation of those bytes.
+	// It returns an error if the conversion fails, or if the functionality is not supported by the app.
+	Convert(ctx context.Context, req ConversionRequest) (*RawObject, error)
+	// CallSubresource handles the subresource call, and writes the response to the http.ResponseWriter.
+	// If a non-http-error response is encountered, an error should be returned.
+	// It returns an error if the functionality is not supported by the app.
+	CallSubresource(ctx context.Context, writer http.ResponseWriter, req *SubresourceRequest) error
```

Comment:
QUESTION – Should we abstract the response handling here? Maybe something like:
```suggestion
	CallSubresource(ctx context.Context, req *SubresourceRequest) (RawObject, error)
```
where `RawObject` could be something like:
```golang
type RawObject interface {
	runtime.Object
	schema.ObjectKind
	metav1.Object
}
```

WDYT?

Reply:
If this is the case then we need to rethink the wording and the approach - the described use case does not fit the subresource bill.

Reply:
I've updated the method name and the response is now a more concrete object, though still not necessarily a kubernetes one.

Reply:
Looks much cleaner, thank you!